docs/notebooks/trust_region.pct.py

```diff
@@ -0,0 +1,283 @@
+
+
+# %% [markdown]
+# ## Trust region `TREGO` acquisition rule
+#
+# First we show how to run Bayesian optimization with the `TREGO` algorithm. This is a trust region
+# algorithm that alternates between regular EGO steps and local steps within one trust region.
+#
+# ### Create `TREGO` rule and run optimization loop
+#
+# We can run the Bayesian optimization loop by defining a `BayesianOptimizer` and calling its
+# `optimize` method with the trust region rule. Once the optimization loop is complete, the
+# optimizer will return one new query point for every step in the loop; that's 5 points in total.
+
+# %%
+acq_rule = trieste.acquisition.rule.TrustRegion()
+bo = trieste.bayesian_optimizer.BayesianOptimizer(observer, search_space)
+
+num_steps = 5
+result = bo.optimize(
+    num_steps, initial_data, build_model(), acq_rule, track_state=True
+)
+dataset = result.try_get_final_dataset()
+
+# %% [markdown]
+# ### Visualizing `TREGO` results
+#
+# Let's take a look at where we queried the observer, the original query points (crosses), new
+# query points (dots) and the optimum point found (purple dot), and where they lie with respect to
+# the contours of the Branin.
+
+# %%
+from trieste.experimental.plotting import plot_bo_points, plot_function_2d
+
+
+def plot_final_result(_dataset: trieste.data.Dataset) -> None:
+    arg_min_idx = tf.squeeze(tf.argmin(_dataset.observations, axis=0))
+    query_points = _dataset.query_points.numpy()
+    _, ax = plot_function_2d(
+        branin,
+        search_space.lower,
+        search_space.upper,
+        grid_density=40,
+        contour=True,
+    )
+
+    plot_bo_points(query_points, ax[0, 0], num_initial_data_points, arg_min_idx)
+
+
+plot_final_result(dataset)
+
+# %% [markdown]
+# We can also visualize the progress of the optimization by plotting the trust regions at each step.
+# The trust regions are shown as translucent boxes, with the current optimum point in each region
+# shown in matching color.
+#
+# Note there is only one trust region in this plot, but the rule in the next section will show multiple trust
+# regions.
+
+# %%
+import base64
+import io
+from typing import List
+
+import imageio
+import IPython
+import matplotlib.pyplot as plt
+
+from trieste.experimental.plotting import plot_trust_region_history_2d
+
+
+def fig_to_frame(fig: plt.Figure) -> np.ndarray:
+    fig.canvas.draw()
+    size_pix = fig.get_size_inches() * fig.dpi
+    image = np.frombuffer(fig.canvas.tostring_rgb(), dtype="uint8")
+    return image.reshape(list(size_pix[::-1].astype(int)) + [3])
+
+
+def frames_to_gif(
+    frames: List[np.ndarray], duration=5000
+) -> IPython.display.HTML:
+    gif_file = io.BytesIO()
+    imageio.mimsave(gif_file, frames, format="gif", loop=0, duration=duration)
+    gif = IPython.display.HTML(
+        '<img src="data:image/gif;base64,{0}"/>'.format(
+            base64.b64encode(gif_file.getvalue()).decode()
+        )
+    )
+    return gif
```

Comment:
this is fairly general code that is likely to be useful again, we should move this to experimental/plotting package?

Reply:
Moving the second function to `plotting.py` will add dependencies to `imageio` and `IPython` to the main trieste requirements. Currently those dependencies are only present for notebooks. We probably don't want to do that, do we?

Reply:
we have those plotting extras at https://github.com/secondmind-labs/trieste/blob/93f44b321e7b9112016288b48325c8863feb173e/setup.py#L52, is this for `experimental/plotting`? then we could add `imageio` to this list? 

Ipython dependency makes less sense but the function could return the `gif_file` instead and final conversion to `gif` could be done within notebook